# Plotread on Compute1 (after having potential denovo call)

### Docker: 
```spashleyfu/plotreads:latest```
```
bsub -Is -G compute-jin810 -q general-interactive -a 'docker(spashleyfu/plotreads:latest)' -n 8 -R "select[mem>64GB] rusage[mem=64GB]" /bin/bash

plotReads [-t] [-b] [-d] [-mq #] [-gt X/X] [-impact LOW|MEDIUM|HIGH] [-denovo] [-w colwidth] [-rw readwidth] [-i prefix] [-r reffile] posOrVcf bamfile
            posOrVcf can be a 'chr:pos' position value, a file of those position values, a VCF file or a vcfAnnotate XLS file.
            -t         - Generate text output, instead of images.
            -b         - In text mode, show bases instead of differences.
            -i prefix  - Prefix for the output image file(s). [def:  <bamfile>_plotreads]
            -r reffile - Fasta of reference [def: use read alignments to infer reference]
            -w #       - Set the number of columns of the output.  [def: 70 in text output, 50 in image].
            -rw #      - Include reads whose alignment overlaps this close to the position. [def: 10 for text, 1 for image].
            -d         - Include PCR duplicate reads in output.
            -mq #      - Filter reads with mapq below threshold.  [default 0]
            -gt X/X    - With VCF file input, only show variants with given genotype.
            -impact LOW|MEDIUM|HIGH - With annotation input, only show variants with that impact level r higher.
            -denovo    - With VCF/Annotation generated by denovoFilter, only show de novo variants.``
```
### Usage:
To generate Image: ``` plotReads -r $REF $CHR:$POS $BAM``` </br>
To generate Text file: ```plotReads -t -b -r $REF $CHR:$POS $BAM > file_name.txt```
- $REF: the pathe for reference file were used for mapping in bam file
```/storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa```
- $CHR:$POS means the location of variants: 
    * The $CHR is the chromosome name, for hg38, $CHR must start with "chr".
    * The $POS is the position of variants in the chromosome.
    * For example, for hg38, the $CHR:$POS should looks like `chr2:171692783`; for hg19, it should looks like `2:171692783`.

### Sample script to generate the plotRead command for Family (Trio)
- Full script:[here](https://github.com/Drdreammaerd/Whole_genome_analysis/blob/main/Downstream%20Analysis/Scripts/Plotread_script_Trio.sh)
```

#!/bin/bash

PROJECT="2021_Jun_CP_WGS_6F"

My_DIR="/storage1/fs1/jin810/Active/yung-chun"
Result_DIR="$My_DIR"/results
Denovo_DIR="$Result_DIR"/$PROJECT/Denovo

Variant_Table="$Denovo_DIR/CP_6trio_triodenovo_variantTable_exome.tsv"

while read line; do

    # Skip first line which is header
    test $n -eq 1 && ((n=n+1)) && continue

    # Get Variables:
    SAMPLENAME=$(echo $line | awk '{print $2}') # ie, KRU-F374-003-A
    CHR=$(echo $line | awk '{print $3}')
    POS=$(echo $line | awk '{print $4}')
    FAMID=$(echo $SAMPLENAME | cut -d'-' -f 2)
    #BAMPATH=$VCF_DIR/$FAMID/$SAMPLENAME/$SAMPLENAME"_germline.bam"   

    # Generate the directory for each family under PlotReads
    FAM_DIR="$PLOT_DIR"/$FAMID
    [ ! -d $FAM_DIR ] && mkdir -p $FAM_DIR

    BAM_lst=( $(ls $VCF_DIR/$FAMID/*/*.bam))

    # Change directory to each family
    echo "cd $FAM_DIR" >> $pb_script

    bamnum=1
    # Generate the command line
    for i in ${BAM_lst[@]};
    do
      BAMPATH=$i
      BAMNAME=$(echo $i | cut -d'/' -f 11)
      echo "##"$candidate":"$bamnum"" >> $pb_script
      echo "plotReads -r $REF38 $CHR:$POS $BAMPATH" >> $pb_script
      echo "" >> $pb_script
      echo "plotReads -t -b -r $REF38 $CHR:$POS $BAMPATH > "$BAMNAME"_germline_plotreads_"$CHR"_"$POS".txt " >> $pb_script
      echo "" >> $pb_script
      ((bamnum+=1))
    done
    ((candidate+=1))
done < $Variant_Table

```

### Output
- Image
  ![Screen Shot 2021-09-14 at 12 01 37 PM](https://user-images.githubusercontent.com/86803803/133301754-9031b5b3-7836-4773-b7e9-a20fb7ceae7d.png)
- Text
  ![Screen Shot 2021-09-14 at 12 00 37 PM](https://user-images.githubusercontent.com/86803803/133301665-38695353-d696-48f8-9c66-9ea9a3ff3cfb.png)
- True call from proband
  ![Screen Shot 2021-09-14 at 12 03 58 PM](https://user-images.githubusercontent.com/86803803/133302043-1599ef93-50d9-4bfd-adae-fd90473ba1cd.png)


### Selection criteria
- Base-pair quality of mutation call should be good (the width of each colour tile, the thicker means higher quality).
- mutation should randomly distribute in both forward and reverse strand.
- Sequce aligment score should be high (Max: 60) --> pick three reads (Top, middle, bottom) to re-align in NCSC genome browser (BLAT) (http://genome.ucsc.edu/cgi-bin/hgBlat) to see if the sequence only hits **"one"** target
- Parent shouldn't have any mutation call from plotread. Besides, the alignment and base-pair score should be good enough.
- If the mutation located in the GC repeated region, should be really careful. There are many tools to exclude the **Complex region**
