# PlotRead on compute1

> Script: `/storage1/fs1/jin810/Active/demo_plotRead/generate_plotread_forCP_basic_example.sh`
> 
> Docker Image: [`sam16711/plot_reads:v1`](https://github.com/jinlab-washu/plot_reads) or `spashleyfu/plotreads:latest`
> 
> compute1: `/storage1/fs1/jin810/Active/demo_plotRead`

Running Docker on compute1:

```
$ bsub -Is -G compute-jin810 -q general-interactive -a 'docker(spashleyfu/plotreads:latest)' -R "select[mem>32000] rusage[mem=32000]" /bin/bash

$ plotReads                                                            
/usr/local/lib/python2.7/dist-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
Usage:  plotReads [-t] [-b] [-d] [-mq #] [-gt X/X] [-impact LOW|MEDIUM|HIGH] [-denovo] [-w colwidth] [-rw readwidth] [-i prefix] [-r reffile] posOrVcf bamfile
            posOrVcf can be a 'chr:pos' position value, a file of those position values, a VCF file or a vcfAnnotate XLS file.
            -t         - Generate text output, instead of images.
            -b         - In text mode, show bases instead of differences.
            -i prefix  - Prefix for the output image file(s). [def:  <bamfile>_plotreads]

            -r reffile - Fasta of reference [def: use read alignments to infer reference]
            -w #       - Set the number of columns of the output.  [def: 70 in text output, 50 in image].
            -rw #      - Include reads whose alignment overlaps this close to the position. [def: 10 for text, 1 for image].

            -d         - Include PCR duplicate reads in output.
            -mq #      - Filter reads with mapq below threshold.  [default 0]
            -gt X/X    - With VCF file input, only show variants with given genotype.
            -impact LOW|MEDIUM|HIGH - With annotation input, only show variants with that impact level r higher.
            -denovo    - With VCF/Annotation generated by denovoFilter, only show de novo variants.

```

#### How to run plotRead?

The basic command for plotRead is `plotReads -r $REF $CHR:$POS $BAM`.

So, all you need to do are having:

1. Absolute path of $REF: The reference file you mapping.
2. All $CHR:$POS of variants: 
    * The $CHR is the chromosome name, for hg38, $CHR must start with "chr".
    * The $POS is the position of variants in the chromosome.
    * For example, for hg38, the $CHR:$POS should looks like `chr2:171692783`; for hg19, it should looks like `2:171692783`.
3. The $BAM is the absolute BAM file path for each sample.

##### Here are an real example using CerebralPasly project

**DEMO folder: `/storage1/fs1/jin810/Active/demo_plotRead`**

That's assume you have a filtered variant table TSV file, `CP_6trio_triodenovo_variantTable_exome_TOY.tsv`. 

Also, your BAM file is under `/storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF` folder, it structured like `VCF/$FMAILY_IS/$SAMPLE_NAME/$SAMPLE_NAME'_germline.bam'`.

So, write a script loop through every variants you want to visualize. The example script below is a bash script, you can also using R/Python to do it.

**/storage1/fs1/jin810/Active/demo_plotRead/generate_plotread_forCP_basic_example.sh**

```bash
#!/bin/bash

TSV="/storage1/fs1/jin810/Active/demo_plotRead/CP_6trio_triodenovo_variantTable_exome_TOY.tsv"
BAM_FOLDER="/storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF"
REF38="/storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa"

i=0
while read line; do
    ((i=i+1))
    if [[ $i -eq 1 ]]
    then 
        continue
    fi
    # Get Variables:
    SAMPLENAME=$(echo $line | awk '{print $2}') # ie, KRU-F374-003-A
    CHR=$(echo $line | awk '{print $3}') # ie, chr2
    POS=$(echo $line | awk '{print $4}') # ie, 171692783
    FAMID=$(echo $SAMPLENAME | cut -d'-' -f 2) # ie, F374
    BAMPATH=$BAM_FOLDER/$FAMID/$SAMPLENAME/$SAMPLENAME"_germline.bam"
    echo "plotReads -r $REF38 $CHR:$POS $BAMPATH"
    # Run the command:
    # After test the commands print out above, you can un-comment the command below:
    #$(plotReads -r $REF $CHR:$POS $BAMPATH)
done < $TSV

```

**Running the script:**

The plotRead program will automatically generate the output file named like: `KRU-F374-003-A_germline_plotreads_chrchr2_171692783.png`

```
fup@compute1-exec-152:/storage1/fs1/jin810/Active/demo_plotRead$ ./generate_plotread_forCP_basic_example.sh 
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr2:171692783 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr4:968291 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr4:108621975 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr4:108621977 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr10:34516978 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr11:1084313 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr14:57865574 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr18:45063519 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr11:95850751 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F371/KRU-F371-002-U/KRU-F371-002-U_germline.bam

// TEST the commands:
fup@compute1-exec-152:/storage1/fs1/jin810/Active/demo_plotRead$ plotReads -r /storage1/fs1/bga/Active/gmsroot/gc2560/core/model_data/2887491634/build21f22873ebe0486c8e6f69c15435aa96/all_sequences.fa chr2:171692783 /storage1/fs1/jin810/Active/yung-chun/results/2021_Jun_CP_WGS_6F/VCF/F374/KRU-F374-003-A/KRU-F374-003-A_germline.bam
Writing chr2:171692783...

fup@compute1-exec-152:/storage1/fs1/jin810/Active/demo_plotRead$ ls
CP_6trio_triodenovo_variantTable_exome_TOY.tsv		 README
generate_plotread_forCP_basic_example.sh
generate_plotread_forCP_advence_example.sh
KRU-F374-003-A_germline_plotreads_chrchr2_171692783.png

```

#### Practices:

Basically, we will visualize the variants Trio by Trio. Can you modify the script to generate all plotRead commands family by family?
